### DISCUSSION TITLE
New to TorchSharp: Need help with torch.nn.Module forward method overriding

### DISCUSSION THREAD

**Discussion by dustincoding** [xamarin/TorchSharp#1215] [Q&A] [answered]
I started one of my very first projects using TorchSharp and converting a well known example of transformer architecture to C# from Python. I've run into a bit of a roadblock while trying to override the forward method for my modules. All of my modules inherit from torch.nn.Module but cannot find any suitable method to override. The project that I created is a simple windows application in .net 8.0 and I am using the nuget package for TorchSharp 0.101.5. I guess I'm just confused on whether I should be implementing a different signature for the Module class or if this is truly a bug. I'd really appreciate any help that anyone can offer to get this project moving again.

`
using TorchSharp.Modules;
using static TorchSharp.torch;
using static TorchSharp.torch.nn;

namespace BigramCommunicationNNModel.Model
{
    public class Block : Module
    {
        private Module multiHeadAttention;
        private Module feedForward;
        private LayerNorm ln1;
        private LayerNorm ln2;

        public Block(int n_embd, int n_head, double dropoutVal, int blockSize): base("Block")
        {
            int headSize = n_embd / n_head;
            multiHeadAttention = new MultiHeadAttention(n_embd, n_head, headSize, dropoutVal, blockSize);
            feedForward = new FeedForward(n_embd, dropoutVal);
            ln1 = LayerNorm(new long[] { n_embd });
            ln2 = LayerNorm(new long[] {n_embd });

        }

        public override Tensor forward(Tensor x)
        {
            Tensor x1 = ln1.forward(x);
            Tensor x2 = ((MultiHeadAttention)multiHeadAttention).forward(x1);
            x = x + x2;
            var x3 = ln2.forward(x);
            var x4 = ((FeedForward)feedForward).forward(x3);
            x = x + x4;

            return x;

        }

        private void RegisterComponents()
        {
            this.register_module("multiHeadAttention", multiHeadAttention);
            this.register_module("feedForward", feedForward);
            this.register_module("ln1", ln1);
            this.register_module("ln2", ln2);
        }
    }
}`

**Comment by asieradzk** [comment, marked as answer]
You need to inherit from Module<Tensor, Tensor>

Then your inheriting class will have forward method to override

```cs
class SomeClass: Module<Tensor, Tensor>
    {
        public SomeClass(string name) : base(name)
        {
        }
```

**Reply by dustincoding** [reply]
Thank you very much this was it. It looks like Module<Tensor, Tensor> is of Module<T, TResult> which has the forward method for overriding.
